classes/deletemodule.php

```diff
@@ -0,0 +1,144 @@
+     *
+     * @param int $taskid The taskid (aka id field for {task_adhoc} table)
+     * @param int $coursemoduleid The course_module record id for the module being deleted.
+     * @param int $moduleinstanceid The moduleinstance id for the module being deleted.
+     * @param int $courseid The course id in which the module being deleted is situated.
+     * @param int $section The section id for the module being deleted.
+     */
+    public function __construct(int $taskid,
+                                int $coursemoduleid,
+                                ?int $moduleinstanceid = null,
+                                ?int $courseid = null,
+                                ?int $section = null) {
+        $this->taskid = $taskid;
+        $this->coursemoduleid = $coursemoduleid;
+        $this->moduleinstanceid = $moduleinstanceid;
+        $this->courseid = $courseid;
+        $this->section = $section;
+        $this->set_contextid($this->coursemoduleid);
+        $this->set_modulename($this->coursemoduleid, $this->moduleinstanceid);
+    }
+
+    /**
+     * get_contextid() - Get the name of the table related to the course module which is failing to be deleted.
+     *
+     * @return string
+     */
+    public function get_contextid() {
+        return $this->modulecontextid;
+    }
+
+
+    /**
+     * get_modulename() - Get the name of the table related to the course module which is failing to be deleted.
+     *
+     * @return string
+     */
+    public function get_modulename() {
+        return $this->modulename;
+    }
+
+    /**
+     * set_contextid() - Set the contextid of the module, retrived from the database record.
+     *
+     * @param int $coursemoduleid - course module instance id.
+     *
+     * @return null
+     */
+    public function set_contextid(int $coursemoduleid) {
+        if (isset($coursemoduleid)) {
+            global $DB;
+
+            if ($result = $DB->get_records('context', array('contextlevel' => '70', 'instanceid' => $coursemoduleid))) {
+                $this->modulecontextid = current($result)->id;
+            } else {
+                $this->modulecontextid = null;
+            }
+        }
+    }
+
+    /**
+     * set_modulename()-  Set the name of the table related to the course module which is failing to be deleted.
+     * Retrived from the database record.
+     *
+     * @param int $coursemoduleid - coursemodule id.
+     * @param int $moduleinstanceid - moduleinstance id.
+     *
+     * @return null
+     */
+    public function set_modulename(int $coursemoduleid, ?int $moduleinstanceid = null) {
+        global $DB;
+        // First get moduleid.
+        if (!isset($moduleinstanceid)) {
+            $queryarray = array('id' => $coursemoduleid);
+        } else {
+            $queryarray = array('id' => $coursemoduleid, 'instance' => $moduleinstanceid);
+        }
+
+        if ($cmrecord = $DB->get_records('course_modules', $queryarray, '', 'module')) {
```

Comment:
This would fail to obtain the modulename when the module's table (e.g. quiz) record was unexpectedly absent. Whereas direct DB query without JOINing to module's table will still get the modulename.
This is due to this query in get_coursemodule_from_id():
https://github.com/moodle/moodle/blob/master/lib/datalib.php#L1257